runtime/runtime_impl.hpp

```diff
@@ -64,21 +67,21 @@ class runtime_impl : public runtime {
 	}
 
 	virtual void wait() override {
-		while (!terminate.load()) {
+		while (!pb.lookup_impl<Stoplight>()->should_stop()) {
 			std::this_thread::sleep_for(std::chrono::milliseconds{10});
 		}
 	}
 
 	virtual void stop() override {
+		pb.lookup_impl<Stoplight>()->stop();
```

Comment:
Setting this before doing the actual stop would unblock `wait()` while the stop is in progress. Move to end of function.

Reply:
`Stoplight::stop()` needs to happen first, and then each `threadloop::stop()`.

`Stoplight::stop()` tells the plugin threads to stop after `_p_one_iteration()`; `threadloop::stop()` joins on the plugin threads (waits for them to terminate).

Reply:
Right, that makes sense. But I was talking about `runtime_impl->wait()`. As soon as you call `pb.lookup_impl<Stoplight>()->stop()`, `runtime_impl->wait()` will return, meaning that [this](https://github.com/ILLIXR/ILLIXR/blob/master/runtime/main.cpp#L98) will unblock and cause an exit possibly before the plugins have actually stopped. Or am I missing something?

Reply:
Confirming: `wait_for_shutdown()` and `shutdown_done()` were added to address this right?

Reply:
Yes.@mhuzai, if you are satisfied by this solution, mark conversation as resolved.